crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -1211,7 +1211,20 @@ impl OlmMachine {
             account.update_key_counts(
                 sync_changes.one_time_keys_counts,
                 sync_changes.unused_fallback_keys,
-            )
+            );
+
+            // Flush the applicative transaction as soon as the keys have been generated. If
+            // we don't do it here, then it's possible that the
+            // `StoreTransaction` fails, but we've uploaded keys in the
```

Comment:
How would it happen? If the store transaction fails, then the `receive_sync_changes` will fail, then the whole sync will fail and the keys wont be uploaded?

I think that in that case the sync token wouldn't be persisted, so the app will try again the same sync?